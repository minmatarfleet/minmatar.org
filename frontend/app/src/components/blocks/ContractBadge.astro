---
import { i18n } from '@helpers/i18n'
const { lang, t } = i18n(Astro.url)

import type { ContractUI } from '@dtypes/layout_components'

interface Props {
    contract:       ContractUI;
    is_supplier:    boolean;
}

const {
    contract,
    is_supplier,
} = Astro.props

const low_amount = (contract.current_quantity <= (20 * contract.desired_quantity)/100)

import { get_item_icon } from '@helpers/eve_image_server'
import { format_date_shortest } from '@helpers/date'

import FlexInline from '@components/compositions/FlexInline.astro';
import Wrapper from '@components/compositions/Wrapper.astro';
import FixedFluid from '@components/compositions/FixedFluid.astro';
import Flexblock from '@components/compositions/Flexblock.astro';

import ComponentBlock from '@components/blocks/ComponentBlock.astro';
import CharacterPicture from '@components/blocks/CharacterPicture.astro';
import CorporationPicture from '@components/blocks/CorporationPicture.astro';
import Square from '@components/blocks/Square.astro';
---

<ComponentBlock padding_inline='0' padding_block='0'>
    <Flexblock gap='0'>
        <Wrapper padding_block='var(--space-2xs)' padding_inline='var(--space-2xs)'>
            <FlexInline justification='space-between'>
                <FixedFluid width='32px' class="[ items-center grow ]">
                    <picture>
                        <img loading="lazy" src={get_item_icon(contract.eve_type_id, 32)} width={32} height={32} />
                    </picture>
                    <h4 class="[ title ]">{contract.title}</h4>
                </FixedFluid>
                <div class="[ quantity ]">
                    <span class:list={{ 'text-red': low_amount }}>{contract.current_quantity}</span>/{contract.desired_quantity}
                </div>
            </FlexInline>
        </Wrapper>
        <div class="h-[64px] mb-[-7px]">
            <canvas
                x-data={`{
                    init() {
                        if (window.innerWidth < 1280) {
                            $el.closest('div').remove()
                            return
                        }

                        const ctx = $el.getContext('2d')

                        const gradient = ctx.createLinearGradient(0, 0, 0, 64)
                        gradient.addColorStop(0, '${low_amount ? 'rgba(255, 110, 110, 0.5)' : 'rgba(110, 233, 255, 0.5)'}')
                        gradient.addColorStop(1, '${low_amount ? 'rgba(255, 110, 110, 0)' : 'rgba(110, 233, 255, 0)'}')

                        Chart.defaults.font.family = 'Montserrat Variable';
                        Chart.defaults.backgroundColor = 'rgba(0, 0, 0, 0)';

                        const myChart = new Chart(ctx, {
                            type: 'line',
                            data: {
                                labels: ${JSON.stringify(contract.trend_x_axis.reverse().map(x => format_date_shortest(lang, x)))},
                                datasets: [{
                                    label: "${t('contracts_sold')}",
                                    data: ${JSON.stringify(contract.trend_y_axis.reverse())},
                                    pointStyle: 'rect',
                                    fill: true,
                                    backgroundColor: gradient,
                                    borderColor: '${low_amount ? '#ff6e6e' : '#6ee9ff'}',
                                    borderWidth: 2,
                                    pointRadius: 1
                                },
                                {
                                    label: "${t('desired_quantity')}",
                                    data: ${JSON.stringify(Array(contract.trend_x_axis.length).fill(contract.desired_quantity))},
                                    pointStyle: 'rect',
                                    backgroundColor: 'rgba(241, 217, 160, 0.2)',
                                    borderColor: 'rgba(241, 217, 160, 0.2)',
                                    borderWidth: 1,
                                    pointRadius: 1
                                }]
                            },
                            options: {
                                responsive: true,
                                maintainAspectRatio: false,
                                layout: {
                                    padding: {
                                        left: -10,
                                        bottom: -10
                                    }
                                },
                                plugins: {
                                    legend: {
                                        display: false
                                    },
                                    glow: {
                                        consoleText: 'testText'
                                    },
                                },
                                scales: {
                                    x: {
                                        title: {
                                            display: false
                                        },
                                        grid: {
                                            display: false
                                        },
                                        ticks: {
                                            display: false
                                        },
                                        border: {
                                            display: false
                                        }
                                    },
                                    y: {
                                        title: {
                                            display: false
                                        },
                                        grid: {
                                            display: false
                                        },
                                        ticks: {
                                            display: false
                                        },
                                        border: {
                                            display: false
                                        },
                                        min: 0,
                                        max: ${Math.max(Math.max(...contract.trend_y_axis), contract.desired_quantity) + 1}
                                    }
                                }
                            },
                            plugins: [{
                                id: 'glow',
                                beforeDraw: (chart, args, options) => {
                                    let ctx = chart.ctx
                                    let _stroke = ctx.stroke
                                    ctx.stroke = function() {
                                        ctx.save()
                                        ctx.shadowColor = '#ffffff'
                                        ctx.shadowBlur = 5
                                        ctx.shadowOffsetX = 0
                                        ctx.shadowOffsetY = 0
                                        _stroke.apply(this, arguments)
                                        ctx.restore()
                                    }
                                }
                            }]
                        })
                    }
                }`}
            ></canvas>
        </div>

        <hr class:list={[ 'border', { 'mb-[var(--space-2xs)]': !is_supplier } ]} />

        {is_supplier &&
            <Wrapper padding_block='var(--space-3xs)' padding_inline='var(--space-2xs)'>
                {contract.responsabilities.length > 0 ?
                    <FlexInline gap='var(--space-3xs)'>
                        <small>{contract.responsabilities.length !== 1 ? t('seeders') : t('seeder')}</small>
                        {contract.responsabilities.map(entity =>
                            <div data-tippy-content={entity.entity_name} x-init="tippy($el, tippy_options)">
                                {entity.entity_type === 'character' ?
                                    <CharacterPicture
                                        character_id={entity.entity_id}
                                        character_name={entity.entity_name}
                                        size={16}
                                        icon_quality={32}
                                    />
                                    :
                                    <CorporationPicture
                                        corporation_id={entity.entity_id}
                                        corporation_name={entity.entity_name}
                                        size={16}
                                        icon_quality={32}
                                    />
                                }
                            </div>
                        )}
                    </FlexInline>
                    :
                    <FlexInline gap='var(--space-3xs)'>
                        <Square color='fleet-red' />
                        <small>{t('no_seeders')}</small>
                    </FlexInline>
                }
            </Wrapper>
        }
    </Flexblock>
</ComponentBlock>

<style lang="scss">
    .quantity {
        font-family: var(--heading-font);
        text-transform: uppercase;
        font-size: var(--step-0);
        font-weight: 400;
        color: var(--highlight);
    }
</style>