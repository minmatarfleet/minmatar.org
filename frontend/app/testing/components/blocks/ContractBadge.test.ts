import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import ContractBadge from "@components/blocks/ContractBadge.astro";

test("ContractBadge defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(ContractBadge, {
    props: {
      contract: {
        current_quantity: 56,
        desired_quantity: 60,
        eve_type_id: 12,
        title: "MOOOOREEEE",
        trend_x_axis: [],
        trend_y_axis: [],
      },
    },
    slots: {
      default: "slotted chick",
    },
  });

  expect(result).toMatchSnapshot();
});
