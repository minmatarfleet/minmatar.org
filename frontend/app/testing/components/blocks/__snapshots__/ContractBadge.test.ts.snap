// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContractBadge defaults 1`] = `
"<div class="wrapper [ card component ] light-transparency" data-astro-cid-heszkxoq="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: 0;--wrapper_padding_inline_end: 0;--wrapper_padding_block: 0;--wrapper_max_width: 100%;">   <div class="flex-block gap-0" data-astro-cid-heszkxoq="true" data-mod style="--flex_block_gap: 0;">  <div class="wrapper" data-astro-cid-heszkxoq="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-2xs);--wrapper_padding_inline_end: var(--space-2xs);--wrapper_padding_block: var(--space-2xs);--wrapper_max_width: 80rem;">  <div class="flex-inline" data-astro-cid-heszkxoq="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: space-between;">  <div class="fixed-fluid [ items-center grow ]" data-astro-cid-heszkxoq="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 32px;--fixed_fluid_breakpoint: 50%;">  <picture data-astro-cid-heszkxoq> <img loading="lazy" src="https://images.evetech.net/types/12/icon?size=32" width="32" height="32" data-astro-cid-heszkxoq> </picture> <h4 class="[ title ]" data-astro-cid-heszkxoq>MOOOOREEEE</h4>  </div>   <div class="[ quantity ]" data-astro-cid-heszkxoq> <span data-astro-cid-heszkxoq>56</span>/60 </div>  </div>    </div>   <div class="h-[64px] mb-[-7px]" data-astro-cid-heszkxoq> <canvas x-data="{
                    init() {
                        const ctx = $el.getContext('2d')

                        const gradient = ctx.createLinearGradient(0, 0, 0, 64)
                        gradient.addColorStop(0, 'rgba(110, 233, 255, 0.5)')
                        gradient.addColorStop(1, 'rgba(110, 233, 255, 0)')

                        Chart.defaults.font.family = 'Montserrat Variable';
                        Chart.defaults.backgroundColor = 'rgba(0, 0, 0, 0)';

                        const myChart = new Chart(ctx, {
                            type: 'line',
                            data: {
                                labels: [],
                                datasets: [{
                                    label: &#34;Monthly sales&#34;,
                                    data: [],
                                    pointStyle: 'rect',
                                    fill: true,
                                    backgroundColor: gradient,
                                    borderColor: '#6ee9ff',
                                    borderWidth: 2,
                                    pointRadius: 1
                                },
                                {
                                    label: &#34;Desired quantity on stock&#34;,
                                    data: [],
                                    pointStyle: 'rect',
                                    backgroundColor: 'rgba(241, 217, 160, 0.2)',
                                    borderColor: 'rgba(241, 217, 160, 0.2)',
                                    borderWidth: 1,
                                    pointRadius: 1
                                }]
                            },
                            options: {
                                responsive: true,
                                maintainAspectRatio: false,
                                layout: {
                                    padding: {
                                        left: -10,
                                        bottom: -10
                                    }
                                },
                                plugins: {
                                    legend: {
                                        display: false
                                    },
                                    glow: {
                                        consoleText: 'testText'
                                    },
                                },
                                scales: {
                                    x: {
                                        title: {
                                            display: false
                                        },
                                        grid: {
                                            display: false
                                        },
                                        ticks: {
                                            display: false
                                        },
                                        border: {
                                            display: false
                                        }
                                    },
                                    y: {
                                        title: {
                                            display: false
                                        },
                                        grid: {
                                            display: false
                                        },
                                        ticks: {
                                            display: false
                                        },
                                        border: {
                                            display: false
                                        },
                                        min: 0,
                                        max: 61
                                    }
                                }
                            },
                            plugins: [{
                                id: 'glow',
                                beforeDraw: (chart, args, options) => {
                                    let ctx = chart.ctx
                                    let _stroke = ctx.stroke
                                    ctx.stroke = function() {
                                        ctx.save()
                                        ctx.shadowColor = '#ffffff'
                                        ctx.shadowBlur = 5
                                        ctx.shadowOffsetX = 0
                                        ctx.shadowOffsetY = 0
                                        _stroke.apply(this, arguments)
                                        ctx.restore()
                                    }
                                }
                            }]
                        })
                    }
                }" data-astro-cid-heszkxoq></canvas> </div> <hr class="border mb-[var(--space-2xs)]" data-astro-cid-heszkxoq>  </div>    </div>   "
`;
